docs/manpages/smbpasswd.5
.\" This manpage has been automatically generated by docbook2man-spec
.\" from a DocBook document.  docbook2man-spec can be found at:
.\" <http://shell.ipoline.com/~elmert/hacks/docbook2X/> 
.\" Please send any bug reports, improvements, comments, patches, 
.\" etc. to Steve Cheng <steve@ggi-project.org>.
.TH "SMBPASSWD" "5" "28 February 2001" "" ""
.SH NAME
smbpasswd \- The Samba encrypted password file
.SH SYNOPSIS
.PP
\fIsmbpasswd\fR
.SH "DESCRIPTION"
.PP
This tool is part of the  Samba <URL:samba.7.html> suite.
.PP
smbpasswd is the Samba encrypted password file. It contains 
the username, Unix user id and the SMB hashed passwords of the 
user, as well as account flag information and the time the 
password was last changed. This file format has been evolving with 
Samba and has had several different formats in the past. 
.SH "FILE FORMAT"
.PP
The format of the smbpasswd file used by Samba 2.2 
is very similar to the familiar Unix \fIpasswd(5)\fR 
file. It is an ASCII file containing one line for each user. Each field 
ithin each line is separated from the next by a colon. Any entry 
beginning with '#' is ignored. The smbpasswd file contains the 
following information for each user: 
.TP
\fBname\fR
This is the user name. It must be a name that 
already exists in the standard UNIX passwd file. 
.TP
\fBuid\fR
This is the UNIX uid. It must match the uid
field for the same user entry in the standard UNIX passwd file. 
If this does not match then Samba will refuse to recognize 
this smbpasswd file entry as being valid for a user. 
.TP
\fBLanman Password Hash\fR
This is the LANMAN hash of the users password, 
encoded as 32 hex digits. The LANMAN hash is created by DES 
encrypting a well known string with the users password as the 
DES key. This is the same password used by Windows 95/98 machines. 
Note that this password hash is regarded as weak as it is
vulnerable to dictionary attacks and if two users choose the 
same password this entry will be identical (i.e. the password 
is not "salted" as the UNIX password is). If the user has a 
null password this field will contain the characters "NO PASSWORD" 
as the start of the hex string. If the hex string is equal to 
32 'X' characters then the users account is marked as 
disabled and the user will not be able to 
log onto the Samba server. 

\fBWARNING !!\fR Note that, due to 
the challenge-response nature of the SMB/CIFS authentication
protocol, anyone with a knowledge of this password hash will 
be able to impersonate the user on the network. For this
reason these hashes are known as \fBplain text 
equivalents\fR and must \fBNOT\fR be made 
available to anyone but the root user. To protect these passwords 
the smbpasswd file is placed in a directory with read and 
traverse access only to the root user and the smbpasswd file 
itself must be set to be read/write only by root, with no
other access. 
.TP
\fBNT Password Hash\fR
This is the Windows NT hash of the users 
password, encoded as 32 hex digits. The Windows NT hash is 
created by taking the users password as represented in 
16-bit, little-endian UNICODE and then applying the MD4 
(internet rfc1321) hashing algorithm to it. 

This password hash is considered more secure than
the Lanman Password Hash as it preserves the case of the 
password and uses a much higher quality hashing algorithm. 
However, it is still the case that if two users choose the same 
password this entry will be identical (i.e. the password is 
not "salted" as the UNIX password is). 

\fBWARNING !!\fR. Note that, due to 
the challenge-response nature of the SMB/CIFS authentication
protocol, anyone with a knowledge of this password hash will 
be able to impersonate the user on the network. For this
reason these hashes are known as \fBplain text 
equivalents\fR and must \fBNOT\fR be made 
available to anyone but the root user. To protect these passwords 
the smbpasswd file is placed in a directory with read and 
traverse access only to the root user and the smbpasswd file 
itself must be set to be read/write only by root, with no
other access. 
.TP
\fBAccount Flags\fR
This section contains flags that describe 
the attributes of the users account. In the Samba 2.2 release 
this field is bracketed by '[' and ']' characters and is always 
13 characters in length (including the '[' and ']' characters).
The contents of this field may be any of the characters.
.RS
.TP 0.2i
\(bu
\fBU\fR - This means 
this is a "User" account, i.e. an ordinary user. Only User 
and Workstation Trust accounts are currently supported 
in the smbpasswd file. 
.TP 0.2i
\(bu
\fBN\fR - This means the
account has no password (the passwords in the fields Lanman 
Password Hash and NT Password Hash are ignored). Note that this 
will only allow users to log on with no password if the \fI null passwords\fR parameter is set in the \fIsmb.conf(5)
\fR <URL:smb.conf.5.html#NULLPASSWORDS> config file. 
.TP 0.2i
\(bu
\fBD\fR - This means the account 
is disabled and no SMB/CIFS logins will be allowed for 
this user. 
.TP 0.2i
\(bu
\fBW\fR - This means this account 
is a "Workstation Trust" account. This kind of account is used 
in the Samba PDC code stream to allow Windows NT Workstations 
and Servers to join a Domain hosted by a Samba PDC. 
.RE
.PP
Other flags may be added as the code is extended in future.
The rest of this field space is filled in with spaces. 
.PP
.TP
\fBLast Change Time\fR
This field consists of the time the account was 
last modified. It consists of the characters 'LCT-' (standing for 
"Last Change Time") followed by a numeric encoding of the UNIX time 
in seconds since the epoch (1970) that the last change was made. 
.PP
All other colon separated fields are ignored at this time.
.PP
.SH "VERSION"
.PP
This man page is correct for version 2.2 of 
the Samba suite.
.SH "SEE ALSO"
.PP
\fBsmbpasswd(8)\fR <URL:smbpasswd.8.html>, 
samba(7) <URL:samba.7.html>, and
the Internet RFC1321 for details on the MD4 algorithm.
.SH "AUTHOR"
.PP
The original Samba software and related utilities 
were created by Andrew Tridgell. Samba is now developed
by the Samba Team as an Open Source project similar 
to the way the Linux kernel is developed.
.PP
The original Samba man pages were written by Karl Auer. 
The man page sources were converted to YODL format (another 
excellent piece of Open Source software, available at
ftp://ftp.icce.rug.nl/pub/unix/ <URL:ftp://ftp.icce.rug.nl/pub/unix/>) and updated for the Samba 2.0 
release by Jeremy Allison. The conversion to DocBook for 
Samba 2.2 was done by Gerald Carter


NEWS
Changes in release 0.3d

 * add krb5-config

 * fix a bug in 3des gss-api mechanism, making it compatible with the
   specification and the MIT implementation

 * make telnetd only allow a specific list of environment variables to
   stop it from setting `sensitive' variables

 * try to use an existing libdes

 * lib/krb5, kdc: use correct usage type for ap-req messages.  This
   should improve compatability with MIT krb5 when using 3DES
   encryption types

 * kdc: fix memory allocation problem

 * update config.guess and config.sub

 * lib/roken: more stuff implemented

 * bug fixes and portability enhancements

Changes in release 0.3c

 * lib/krb5: memory caches now support the resolve operation

 * appl/login: set PATH to some sane default

 * kadmind: handle several realms

 * bug fixes (including memory leaks)

Changes in release 0.3b

 * kdc: prefer default-salted keys on v5 requests

 * kdc: lowercase hostnames in v4 mode

 * hprop: handle more types of MIT salts

 * lib/krb5: fix memory leak

 * bug fixes

Changes in release 0.3a:

 * implement arcfour-hmac-md5 to interoperate with W2K

 * modularise the handling of the master key, and allow for other
   encryption types. This makes it easier to import a database from
   some other source without having to re-encrypt all keys.

 * allow for better control over which encryption types are created

 * make kinit fallback to v4 if given a v4 KDC

 * make klist work better with v4 and v5, and add some more MIT
   compatibility options

 * make the kdc listen on the krb524 (4444) port for compatibility
   with MIT krb5 clients

 * implement more DCE/DFS support, enabled with --enable-dce, see
   lib/kdfs and appl/dceutils

 * make the sequence numbers work correctly

 * bug fixes

Changes in release 0.2t:

 * bug fixes

Changes in release 0.2s:

 * add OpenLDAP support in hdb

 * login will get v4 tickets when it receives forwarded tickets

 * xnlock supports both v5 and v4

 * repair source routing for telnet

 * fix building problems with krb4 (krb_mk_req)

 * bug fixes

Changes in release 0.2r:

 * fix realloc memory corruption bug in kdc

 * `add --key' and `cpw --key' in kadmin

 * klist supports listing v4 tickets

 * update config.guess and config.sub

 * make v4 -> v5 principal name conversion more robust

 * support for anonymous tickets

 * new man-pages

 * telnetd: do not negotiate KERBEROS5 authentication if there's no keytab.

 * use and set expiration and not password expiration when dumping
   to/from ka server databases / krb4 databases

 * make the code happier with 64-bit time_t

 * follow RFC2782 and by default do not look for non-underscore SRV names

Changes in release 0.2q:

 * bug fix in tcp-handling in kdc

 * bug fix in expand_hostname

Changes in release 0.2p:

 * bug fix in `kadmin load/merge'

 * bug fix in krb5_parse_address

Changes in release 0.2o:

 * gss_{import,export}_sec_context added to libgssapi

 * new option --addresses to kdc (for listening on an explicit set of
   addresses)

 * bug fixes in the krb4 and kaserver emulation part of the kdc

 * other bug fixes

Changes in release 0.2n:

 * more robust parsing of dump files in kadmin
 * changed default timestamp format for log messages to extended ISO
   8601 format (Y-M-DTH:M:S)
 * changed md4/md5/sha1 APIes to be de-facto `standard'
 * always make hostname into lower-case before creating principal
 * small bits of more MIT-compatability
 * bug fixes

Changes in release 0.2m:

 * handle glibc's getaddrinfo() that returns several ai_canonname

 * new endian test

 * man pages fixes

Changes in release 0.2l:

 * bug fixes

Changes in release 0.2k:

 * better IPv6 test

 * make struct sockaddr_storage in roken work better on alphas

 * some missing [hn]to[hn]s fixed.

 * allow users to change their own passwords with kadmin (with initial
   tickets)

 * fix stupid bug in parsing KDC specification

 * add `ktutil change' and `ktutil purge'

Changes in release 0.2j:

 * builds on Irix

 * ftpd works in passive mode

 * should build on cygwin

 * work around broken IPv6-code on OpenBSD 2.6, also add configure
   option --disable-ipv6

Changes in release 0.2i:

 * use getaddrinfo in the missing places.

 * fix SRV lookup for admin server

 * use get{addr,name}info everywhere.  and implement it in terms of
   getipnodeby{name,addr} (which uses gethostbyname{,2} and
   gethostbyaddr)

Changes in release 0.2h:

 * fix typo in kx (now compiles)

Changes in release 0.2g:

 * lots of bug fixes:
   * push works
   * repair appl/test programs
   * sockaddr_storage works on solaris (alignment issues)
   * works better with non-roken getaddrinfo
   * rsh works
   * some non standard C constructs removed

Changes in release 0.2f:

 * support SRV records for kpasswd
 * look for both _kerberos and krb5-realm when doing host -> realm mapping

Changes in release 0.2e:

 * changed copyright notices to remove `advertising'-clause.
 * get{addr,name}info added to roken and used in the other code
   (this makes things work much better with hosts with both v4 and v6
    addresses, among other things)
 * do pre-auth for both password and key-based get_in_tkt
 * support for having several databases
 * new command `del_enctype' in kadmin
 * strptime (and new strftime) add to roken
 * more paranoia about finding libdb
 * bug fixes

Changes in release 0.2d:

 * new configuration option [libdefaults]default_etypes_des
 * internal ls in ftpd builds without KRB4
 * kx/rsh/push/pop_debug tries v5 and v4 consistenly
 * build bug fixes
 * other bug fixes

Changes in release 0.2c:

 * bug fixes (see ChangeLog's for details)

Changes in release 0.2b:

 * bug fixes
 * actually bump shared library versions

Changes in release 0.2a:

 * a new program verify_krb5_conf for checking your /etc/krb5.conf
 * add 3DES keys when changing password
 * support null keys in database
 * support multiple local realms
 * implement a keytab backend for AFS KeyFile's

source/utils/smbpasswd.c
#ifdef SMB_PASSWD

/*
 * Unix SMB/Netbios implementation. Version 1.9. smbpasswd module. Copyright
 * (C) Jeremy Allison 1995.
 * 
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License, or (at your option)
 * any later version.
 * 
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
 * more details.
 * 
 * You should have received a copy of the GNU General Public License along with
 * this program; if not, write to the Free Software Foundation, Inc., 675
 * Mass Ave, Cambridge, MA 02139, USA.
 */

#include "includes.h"
#include "des.h"

/* Static buffers we will return. */
static struct smb_passwd pw_buf;
static pstring  user_name;
static unsigned char smbpwd[16];
static unsigned char smbntpwd[16];

static int gethexpwd(char *p, char *pwd)
{
	int i;
	unsigned char   lonybble, hinybble;
	char           *hexchars = "0123456789ABCDEF";
	char           *p1, *p2;
	for (i = 0; i < 32; i += 2) {
		hinybble = toupper(p[i]);
		lonybble = toupper(p[i + 1]);

		p1 = strchr(hexchars, hinybble);
		p2 = strchr(hexchars, lonybble);
		if (!p1 || !p2)
			return (False);

		hinybble = PTR_DIFF(p1, hexchars);
		lonybble = PTR_DIFF(p2, hexchars);

		pwd[i / 2] = (hinybble << 4) | lonybble;
	}
	return (True);
}

struct smb_passwd *
_my_get_smbpwnam(FILE * fp, char *name, BOOL * valid_old_pwd, 
		BOOL *got_valid_nt_entry, long *pwd_seekpos)
{
	char            linebuf[256];
	unsigned char   c;
	unsigned char  *p;
	long            uidval;
	long            linebuf_len;

	/*
	 * Scan the file, a line at a time and check if the name matches.
	 */
	while (!feof(fp)) {
		linebuf[0] = '\0';
		*pwd_seekpos = ftell(fp);

		fgets(linebuf, 256, fp);
		if (ferror(fp))
			return NULL;

		/*
		 * Check if the string is terminated with a newline - if not
		 * then we must keep reading and discard until we get one.
		 */
		linebuf_len = strlen(linebuf);
		if (linebuf[linebuf_len - 1] != '\n') {
			c = '\0';
			while (!ferror(fp) && !feof(fp)) {
				c = fgetc(fp);
				if (c == '\n')
					break;
			}
		} else
			linebuf[linebuf_len - 1] = '\0';

		if ((linebuf[0] == 0) && feof(fp))
			break;
		/*
		 * The line we have should be of the form :-
		 * 
		 * username:uid:[32hex bytes]:....other flags presently
		 * ignored....
		 * 
		 * or,
		 * 
		 * username:uid:[32hex bytes]:[32hex bytes]:....ignored....
		 * 
		 * if Windows NT compatible passwords are also present.
		 */

		if (linebuf[0] == '#' || linebuf[0] == '\0')
			continue;
		p = (unsigned char *) strchr(linebuf, ':');
		if (p == NULL)
			continue;
		/*
		 * As 256 is shorter than a pstring we don't need to check
		 * length here - if this ever changes....
		 */
		strncpy(user_name, linebuf, PTR_DIFF(p, linebuf));
		user_name[PTR_DIFF(p, linebuf)] = '\0';
		if (!strequal(user_name, name))
			continue;

		/* User name matches - get uid and password */
		p++;		/* Go past ':' */
		if (!isdigit(*p))
			return (False);

		uidval = atoi((char *) p);
		while (*p && isdigit(*p))
			p++;

		if (*p != ':')
			return (False);

		/*
		 * Now get the password value - this should be 32 hex digits
		 * which are the ascii representations of a 16 byte string.
		 * Get two at a time and put them into the password.
		 */
		p++;
		*pwd_seekpos += PTR_DIFF(p, linebuf);	/* Save exact position
							 * of passwd in file -
							 * this is used by
							 * smbpasswd.c */
		if (*p == '*' || *p == 'X') {
			/* Password deliberately invalid - end here. */
			*valid_old_pwd = False;
			*got_valid_nt_entry = False;
			pw_buf.smb_nt_passwd = NULL;	/* No NT password (yet)*/

			/* Now check if the NT compatible password is
			   available. */
			p += 33; /* Move to the first character of the line after 
						the lanman password. */
			if ((linebuf_len >= (PTR_DIFF(p, linebuf) + 33)) && (p[32] == ':')) {
				/* NT Entry was valid - even if 'X' or '*', can be overwritten */
				*got_valid_nt_entry = True;
				if (*p != '*' && *p != 'X') {
				  if (gethexpwd((char *)p,(char *)smbntpwd))
				    pw_buf.smb_nt_passwd = smbntpwd;
				}
			}
			pw_buf.smb_name = user_name;
			pw_buf.smb_userid = uidval;
			pw_buf.smb_passwd = NULL;	/* No password */
			return (&pw_buf);
		}
		if (linebuf_len < (PTR_DIFF(p, linebuf) + 33))
			return (False);

		if (p[32] != ':')
			return (False);

		if (!strncasecmp((char *)p, "NO PASSWORD", 11)) {
		  pw_buf.smb_passwd = NULL;	/* No password */
		} else {
		  if(!gethexpwd((char *)p,(char *)smbpwd))
		    return False;
		  pw_buf.smb_passwd = smbpwd;
		}

		pw_buf.smb_name = user_name;
		pw_buf.smb_userid = uidval;
		pw_buf.smb_nt_passwd = NULL;
		*got_valid_nt_entry = False;
		*valid_old_pwd = True;

		/* Now check if the NT compatible password is
		   available. */
		p += 33; /* Move to the first character of the line after 
					the lanman password. */
		if ((linebuf_len >= (PTR_DIFF(p, linebuf) + 33)) && (p[32] == ':')) {
			/* NT Entry was valid - even if 'X' or '*', can be overwritten */
			*got_valid_nt_entry = True;
			if (*p != '*' && *p != 'X') {
			  if (gethexpwd((char *)p,(char *)smbntpwd))
			    pw_buf.smb_nt_passwd = smbntpwd;
			}
		}
		return &pw_buf;
	}
	return NULL;
}

/*
 * Print command usage on stderr and die.
 */
void 
usage(char *name)
{
	fprintf(stderr, "Usage is : %s [username]\n", name);
	exit(1);
}

int main(int argc, char **argv)
{
  int             real_uid;
  struct passwd  *pwd;
  fstring         old_passwd;
  uchar           old_p16[16];
  uchar           old_nt_p16[16];
  fstring         new_passwd;
  uchar           new_p16[16];
  uchar           new_nt_p16[16];
  char           *p;
  struct smb_passwd *smb_pwent;
  FILE           *fp;
  BOOL            valid_old_pwd = False;
  BOOL 			got_valid_nt_entry = False;
  long            seekpos;
  int             pwfd;
  char            ascii_p16[66];
  char            c;
  int             ret, i, err, writelen;
  int             lockfd = -1;
  char           *pfile = SMB_PASSWD_FILE;
  char            readbuf[16 * 1024];
  
  TimeInit();

  setup_logging(argv[0],True);
  
  charset_initialise();
  
#ifndef DEBUG_PASSWORD
  /* Check the effective uid */
  if (geteuid() != 0) {
    fprintf(stderr, "%s: Must be setuid root.\n", argv[0]);
    exit(1);
  }
#endif
  
  /* Get the real uid */
  real_uid = getuid();
  
  /* Deal with usage problems */
  if (real_uid == 0) {
    /* As root we can change anothers password. */
    if (argc != 1 && argc != 2)
      usage(argv[0]);
  } else if (argc != 1)
    usage(argv[0]);
  
  
  if (real_uid == 0 && argc == 2) {
    /* If we are root we can change anothers password. */
    strncpy(user_name, argv[1], sizeof(user_name) - 1);
    user_name[sizeof(user_name) - 1] = '\0';
    pwd = getpwnam(user_name);
  } else {
    pwd = getpwuid(real_uid);
  }
  
  if (pwd == 0) {
    fprintf(stderr, "%s: Unable to get UNIX password entry for user.\n", argv[0]);
    exit(1);
  }
  /* If we are root we don't ask for the old password. */
  old_passwd[0] = '\0';
  if (real_uid != 0) {
    p = getpass("Old SMB password:");
    strncpy(old_passwd, p, sizeof(fstring));
    old_passwd[sizeof(fstring)-1] = '\0';
  }
  new_passwd[0] = '\0';
  p = getpass("New SMB password:");
  strncpy(new_passwd, p, sizeof(fstring));
  new_passwd[sizeof(fstring)-1] = '\0';
  p = getpass("Retype new SMB password:");
  if (strcmp(p, new_passwd)) {
    fprintf(stderr, "%s: Mismatch - password unchanged.\n", argv[0]);
    exit(1);
  }
  
  if (new_passwd[0] == '\0') {
    printf("Password not set\n");
    exit(0);
  }
  
  /* Calculate the MD4 hash (NT compatible) of the old and new passwords */
  memset(old_nt_p16, '\0', 16);
  E_md4hash((uchar *)old_passwd, old_nt_p16);
  
  memset(new_nt_p16, '\0', 16);
  E_md4hash((uchar *) new_passwd, new_nt_p16);
  
  /* Mangle the passwords into Lanman format */
  old_passwd[14] = '\0';
  strupper(old_passwd);
  new_passwd[14] = '\0';
  strupper(new_passwd);
  
  /*
   * Calculate the SMB (lanman) hash functions of both old and new passwords.
   */
  
  memset(old_p16, '\0', 16);
  E_P16((uchar *) old_passwd, old_p16);
  
  memset(new_p16, '\0', 16);
  E_P16((uchar *) new_passwd, new_p16);
  
  /*
   * Open the smbpaswd file XXXX - we need to parse smb.conf to get the
   * filename
   */
  if ((fp = fopen(pfile, "r+")) == NULL) {
    err = errno;
    fprintf(stderr, "%s: Failed to open password file %s.\n",
	    argv[0], pfile);
    errno = err;
    perror(argv[0]);
    exit(err);
  }
  /* Set read buffer to 16k for effiecient reads */
  setvbuf(fp, readbuf, _IOFBF, sizeof(readbuf));
  
  /* make sure it is only rw by the owner */
  chmod(pfile, 0600);
  
  /* Lock the smbpasswd file for write. */
  if ((lockfd = pw_file_lock(pfile, F_WRLCK, 5)) < 0) {
    err = errno;
    fprintf(stderr, "%s: Failed to lock password file %s.\n",
	    argv[0], pfile);
    fclose(fp);
    errno = err;
    perror(argv[0]);
    exit(err);
  }
  /* Get the smb passwd entry for this user */
  smb_pwent = _my_get_smbpwnam(fp, pwd->pw_name, &valid_old_pwd, 
			       &got_valid_nt_entry, &seekpos);
  if (smb_pwent == NULL) {
    fprintf(stderr, "%s: Failed to find entry for user %s in file %s.\n",
	    argv[0], pwd->pw_name, pfile);
    fclose(fp);
    pw_file_unlock(lockfd);
    exit(1);
  }
  /* If we are root we don't need to check the old password. */
  if (real_uid != 0) {
    if ((valid_old_pwd == False) || (smb_pwent->smb_passwd == NULL)) {
      fprintf(stderr, "%s: User %s is disabled, plase contact your administrator to enable it.\n", argv[0], pwd->pw_name);
      fclose(fp);
      pw_file_unlock(lockfd);
      exit(1);
    }
    /* Check the old Lanman password */
    if (memcmp(old_p16, smb_pwent->smb_passwd, 16)) {
      fprintf(stderr, "%s: Couldn't change password.\n", argv[0]);
      fclose(fp);
      pw_file_unlock(lockfd);
      exit(1);
    }
    /* Check the NT password if it exists */
    if (smb_pwent->smb_nt_passwd != NULL) {
      if (memcmp(old_nt_p16, smb_pwent->smb_nt_passwd, 16)) {
	fprintf(stderr, "%s: Couldn't change password.\n", argv[0]);
	fclose(fp);
	pw_file_unlock(lockfd);
	exit(1);
      }
    }
  }
  /*
   * If we get here either we were root or the old password checked out
   * ok.
   */
  /* Create the 32 byte representation of the new p16 */
  for (i = 0; i < 16; i++) {
    sprintf(&ascii_p16[i * 2], "%02X", (uchar) new_p16[i]);
  }
  if(got_valid_nt_entry) {
    /* Add on the NT md4 hash */
    ascii_p16[32] = ':';
    for (i = 0; i < 16; i++) {
      sprintf(&ascii_p16[(i * 2)+33], "%02X", (uchar) new_nt_p16[i]);
    }
  }
  /*
   * Do an atomic write into the file at the position defined by
   * seekpos.
   */
  pwfd = fileno(fp);
  ret = lseek(pwfd, seekpos - 1, SEEK_SET);
  if (ret != seekpos - 1) {
    err = errno;
    fprintf(stderr, "%s: seek fail on file %s.\n",
	    argv[0], pfile);
    fclose(fp);
    errno = err;
    perror(argv[0]);
    pw_file_unlock(lockfd);
    exit(1);
  }
  /* Sanity check - ensure the character is a ':' */
  if (read(pwfd, &c, 1) != 1) {
    err = errno;
    fprintf(stderr, "%s: read fail on file %s.\n",
	    argv[0], pfile);
    fclose(fp);
    errno = err;
    perror(argv[0]);
    pw_file_unlock(lockfd);
    exit(1);
  }
  if (c != ':') {
    fprintf(stderr, "%s: sanity check on passwd file %s failed.\n",
	    argv[0], pfile);
    fclose(fp);
    pw_file_unlock(lockfd);
    exit(1);
  }
  writelen = (got_valid_nt_entry) ? 65 : 32;
  if (write(pwfd, ascii_p16, writelen) != writelen) {
    err = errno;
    fprintf(stderr, "%s: write fail in file %s.\n",
	    argv[0], pfile);
    fclose(fp);
    errno = err;
    perror(argv[0]);
    pw_file_unlock(lockfd);
    exit(err);
  }
  fclose(fp);
  pw_file_unlock(lockfd);
  printf("Password changed\n");
  return 0;
}

#else

#include "includes.h"

int 
main(int argc, char **argv)
{
  printf("smb password encryption not selected in Makefile\n");
  return 0;
}
#endif
